system/util.rs

```diff
@@ -0,0 +1,212 @@
+    use query::query_manager::QueryManager;
+    use std::collections::HashMap;
+    use std::sync::Arc;
+    use storage::durability_client::WALClient;
+    use storage::snapshot::WriteSnapshot;
+    use typeql::query::Pipeline;
+
+    pub fn execute_read_pipeline(
+        tx: TransactionRead<WALClient>, pipeline: &Pipeline
+    ) -> (TransactionRead<WALClient>, Result<Vec<HashMap<String, VariableValue<'static>>>, ()>) {
+        let prepared_pipeline = QueryManager {}.prepare_read_pipeline(
+            tx.snapshot.clone(),
+            &tx.type_manager,
+            tx.thing_manager.clone(),
+            &tx.function_manager,
+            &pipeline,
+        )
+            .unwrap();
+
+        let named_outputs = prepared_pipeline.rows_positions().unwrap().clone();
+
+        let result_as_batch = match prepared_pipeline.into_rows_iterator(ExecutionInterrupt::new_uninterruptible()) {
+            Ok((iterator, _)) => iterator.collect_owned(),
+            Err((err, _)) => {
+                return (tx, Err(()));
+            }
+        };
+
+        match result_as_batch {
+            Ok(batch) => (tx, Ok(collect_answer(batch, named_outputs))),
+            Err(typedb_source) => (tx, Err(())),
+        }
+    }
+
+    pub fn execute_write_pipeline(
+        snapshot: WriteSnapshot<WALClient>,
+        type_manager: &TypeManager,
+        thing_manager: Arc<ThingManager>,
+        function_manager: &FunctionManager,
+        pipeline: &Pipeline
+    ) -> (
+        Result<Vec<HashMap<String, VariableValue<'static>>>, ()>,
+        Arc<WriteSnapshot<WALClient>>
+    ){
+        let prepared_pipeline = QueryManager::new()
+            .prepare_write_pipeline(
+                snapshot,
+                type_manager,
+                thing_manager,
+                function_manager,
+                pipeline,
+            )
+            .unwrap();
+
+        let named_outputs = prepared_pipeline.rows_positions().unwrap().clone();
+
+        let (result_as_batch, snapshot) =
+            match prepared_pipeline.into_rows_iterator(ExecutionInterrupt::new_uninterruptible()) {
+                Ok((iterator, ExecutionContext { snapshot, .. })) => (iterator.collect_owned(), snapshot),
+                Err((err, ExecutionContext { snapshot, .. })) => {
+                    return (Err(()), snapshot);
+                }
+            };
+
+        match result_as_batch {
+            Ok(batch) => (Ok(collect_answer(batch, named_outputs)), snapshot),
+            Err(typedb_source) => (
+                Err(()),
+                snapshot,
+            ),
+        }
+    }
+}
+
+pub mod answer_util {
+    use answer::variable_value::VariableValue;
+    use compiler::VariablePosition;
+    use executor::batch::Batch;
+    use lending_iterator::LendingIterator;
+    use std::collections::HashMap;
+    use database::transaction::TransactionRead;
+    use storage::durability_client::WALClient;
+
+    pub fn collect_answer(
```

Comment:
Really sounds like something possible to be outside of "utils".

Reply:
Currently it's only used by the system package, hence why I'm keeping it there. We can move it when it's needed elsewhere.